### DISCUSSION TITLE
using alembic migration

### DISCUSSION THREAD

**Discussion by nbraun-wolf** [encode/orm#139] [Q&A]
Hi, I am having a lot of trouble using the databases package with SQL alchemy, so I am looking for alternatives. Is it possible to use this ORM with [alembic](https://alembic.sqlalchemy.org/en/latest/)?

Edit: 
Sorry, I haven't seen this sentence in the README. So, seems like it's possible. Are there any examples how to do that?
> Because ORM is built on SQLAlchemy core, you can use Alembic to provide database migrations.

**Comment by aminalaee** [comment]
Yes that's possible, but it's a bit complicated for now, we need to fix it:

You run `alembic init` obviously.
You need to specify `target_metadata` in alembic `env.py` so in your models you have:

```python
# models.py
...
class Product(orm.Model):
    tablename = "products"
    registry = models
    fields = {
        "id": orm.Integer(primary_key=True),
        "name": orm.String(max_length=100),
        "rating": orm.Integer(minimum=1, maximum=5),
        "in_stock": orm.Boolean(default=False),
    }

metadata = models.metadata
```

And for alembic:
```python
# env.py
...
from models import metadata
target_metadata = metadata
```

And then running `alembic revision --autogenerate` works.

**Reply by Kokoserver** [reply]
i tried it, id does not work for me

**Reply by aminalaee** [reply]
You need to provide more information so that people can help. Please provide some information on how it's not working, are you seeing errors?

**Comment by aminalaee** [comment]
And if you're interested to contribute:

We need to change this:
https://github.com/encode/orm/blob/c6b9c0e7a2fee25509a75d1983f203a334376e95/orm/models.py#L28

To:
```python
self._metadata = sqlalchemy.MetaData() 
```

And add a metadata property:
```python
@property
def metadata():
    for model_cls in self.models.values():
            model_cls.build_table()
    return metadata
```

We need this in a few other places too. If you have any feedback or would like to contribute, feel free.

**Reply by bluebrown** [reply]
OK, thanks. I have forked the project, and I am checking it out.

**Reply by aminalaee** [reply]
If you need any help, feel free to ask here.
